fuzzers/run_fuzzer.py

```diff
@@ -0,0 +1,528 @@
+
+    def _output(self, time_log, msg, args, kw):
+        running_for = time_log - self.time_start
+        msg = msg.format(*args, **kw)
+
+        log_prefix = "{:s} - {:s} - {:>5s}: ".format(
+            time_log.isoformat(),
+            self.fuzzer,
+            pretty_timedelta_str(running_for),
+        )
+
+        msg = "\n".join(log_prefix + x for x in msg.splitlines())
+        print(msg, flush=True)
+
+
+def last_lines(f, num, blocksize=100):
+    r"""Return n last lines in a file.
+
+    >>> f = io.StringIO("")
+    >>> last_lines(f, 100)
+    ['']
+
+    >>> f = io.StringIO("1\n2")
+    >>> last_lines(f, 100)
+    ['1', '2']
+
+    >>> f = io.StringIO("11\n22\n33\n44\n55")
+    >>> last_lines(f, 100)
+    ['11', '22', '33', '44', '55']
+
+    >>> f = io.StringIO("11\n22\n33\n44\n55")
+    >>> last_lines(f, 100, 2)
+    ['11', '22', '33', '44', '55']
+
+    >>> f = io.StringIO("11\n22\n33\n44\n55")
+    >>> last_lines(f, 100, 3)
+    ['11', '22', '33', '44', '55']
+
+    >>> f = io.StringIO("11\n22\n33\n44\n55")
+    >>> last_lines(f, 2, 2)
+    ['44', '55']
+
+    >>> f = io.StringIO("11\n22\n33\n44\n55")
+    >>> last_lines(f, 2, 3)
+    ['44', '55']
+
+    >>> f = io.StringIO("11\n22\n33\n44\n55")
+    >>> last_lines(f, 2, 4)
+    ['44', '55']
+
+    >>> f = io.StringIO("1\n22\n3333333\n")
+    >>> last_lines(f, 100, 3)
+    ['1', '22', '3333333', '']
+
+    """
+    lines = collections.deque([''])
+    # Seek to the end of the file
+    f.seek(0, os.SEEK_END)
+    fpos = f.tell()
+
+    while len(lines) < (num + 1) and fpos > 0:
+        bs = min(blocksize, fpos)
+        fpos = fpos - bs
+        f.seek(fpos)
+        data = f.read(bs)
+
+        while True:
+            lpos = data.rfind('\n')
+            lines[0] = data[lpos + 1:] + lines[0]
+            if lpos == -1:
+                break
+            lines.insert(0, '')
+            data = data[:lpos]
+
+    if fpos == 0:
+        lines.insert(0, '')
+
+    return list(lines)[1:]
+
+
+def get_usage():
+    # This function only works if you have a signal handler for the
+    # signal.SIGCHLD signal.
+    raw_usage = resource.getrusage(resource.RUSAGE_CHILDREN)
+    # 0   ru_utime    time in user mode (float)
+    # 1   ru_stime    time in system mode (float)
+    # 2   ru_maxrss   maximum resident set size
+    #
+    # These fields are always zero on Linux
+    # 3   ru_ixrss    shared memory size
+    # 4   ru_idrss    unshared memory size
+    # 5   ru_isrss    unshared stack size
+    return "User:{}s System:{}s".format(
+        int(raw_usage.ru_utime),
+        int(raw_usage.ru_stime),
+    )
+
+
+def get_load():
+    """Return current load average.
+
+    Values is average perecntage of CPU used over 1 minute, 5 minutes, 15
+    minutes.
+    """
+    a1, a5, a15 = os.getloadavg()
+    cpus = os.cpu_count()
+    return a1 / cpus * 100.0, a5 / cpus * 100.0, a15 / cpus * 100.0
+
+
+class PsTree:
+    _pstree_features = None
```

Comment:
Why would you do it this way instead of regular methods and an instance object?

Reply:
This is a pretty normal way to do a singleton style object in Python. 

Making this class needing to be instantiated adds a whole extra level of complexity.

Reply:
But it isn't a singleton, and there is no need to it to be a singleton.

Reply:
The pstree feature detection should only run once and is being cached. Caches like that are generally singletons.

Reply:
Consider using functools.lru_cache of manual caching then?  This is more common and is thread safe in python 3.6 and above.  Alternatively make it an instance class, which makes clear that there is instance data.

Reply:
lru_cache is not really designed for making a single call have permanent results;
> Decorator to wrap a function with a memoizing callable that saves up to the maxsize most recent calls. It can save time when an expensive or I/O bound function is periodically called with the same arguments.

I guess Python really should have something in this category included but it does not seem to...

Reply:
@lru_cache(maxsize=None) should work here.

Reply:
```
import functools

@functools.lru_cache(maxsize=None)
def func():
    print('test')
    return 5

for _ in range(5):
    print(func())
```

Prints:
```
test
5
5
5
5
5
```